### ISSUE TITLE
why there is no "tuio" input source ?

### ISSUE THREAD

**Issue by seeacloud** [TouchScript/TouchScript#309]
when I trying to add tuio input source by:  Add Component -->TouchScript --> input source, but, there have only "standard input", no "tuio" input source, How can I add tuio input source?

I have already checked the "enable TUIO" in the touchscript setting panel, but it seems nothing happened, so sad! need help!

thanks !

**Comment by KevinQ-BIT** [comment]
Hi, seeacloud, I've got the same issue, to solve this problem I just find the file 'TuioInput.cs' and delete the code `#if TOUCHSCRIPT_TUIO`. And don't forget to delete the `#endif` too. This solution is a little bit violent but works fine for me.

**Comment by JosephOHagan** [comment]
@KevinQ-BIT I encountered the same issue though have yet to try your solution. 

Should it just appear as other input sources do? 

Can't test it out until I get into work tomorrow morning (TUIO device is a large touchscreen sphere).

**Comment by KevinQ-BIT** [comment]
@JosephOHagan Yes, it will appear as a input source like the standard one. You can find it by click 'Add Component'.

**Comment by JosephOHagan** [comment]
@KevinQ-BIT Hello, yeah it appears up within the input source list but I don't think it's not picking up any input. I might not be configuring it correctly. Port number is correct. 

How are you using it / do you know how I can get the raw input data?

I basically need to map TUIO input from touching this sphere onto a sphere in Unity. 

Thanks

**Comment by KevinQ-BIT** [comment]
@JosephOHagan Hi, I also had a problem with the configuration as described in issue #312. But if you want to get the raw input data, you can go to file TuioInput.cs line 374. If you have read the official demo of TUIO11 on https://github.com/mkalten/TUIO11_CPP before, you will know how to deal with it. It's a C++ project, but the usage of the functions are the same.

**Comment by JosephOHagan** [comment]
@KevinQ-BIT Thanks for being so helpful.

I'm not sure if this is the same issue but it might just be misunderstanding how to setup TUIO input source. How is it you actually setup the TUIO input source?

I'm not using it with the TUIO simulator rather the actual TUIO hardware when testing. 

So I've setup an empty gameobject and then attach the TUIO input source but then what?

Many thanks in advance (your being a big help)

**Comment by KevinQ-BIT** [comment]
@JosephOHagan Hi, I think you should first read the documents of TouchScript carefully. They showed how to include the TUIO input source. You'll need the TouchManager and Cursors prefab before you use TuioInput. And the structure of the whole project should be like this.
![image](https://user-images.githubusercontent.com/8272436/44973436-2920ce00-af8f-11e8-8a8a-2375d25fc284.png)
Besides, as for raw touch points input, there is also an official example, RawInput, hope it will help you.

**Comment by JosephOHagan** [comment]
@KevinQ-BIT Thanks, yeah I've not had a chance to work on it for a few days (getting back to it today). My structure like yours and from what I can tell from debug statements it connects and checks for TUIO inputs correctly, it just doesn't register the actual inputs when I touch the device. 

I think I'm either overlooking something or it's not configured properly somewhere.

Thanks again though.

**Comment by JosephOHagan** [comment]
@KevinQ-BIT Resolved the issue, it was a setup problem with my device and how it talks to Unity.

Turns out that the TUIO server wasn't connecting properly with the input device itself and I had to reconfig the device itself and handle all of the server stuff within Unity itself. 

UPDATE: Sometimes Windows firewall might block the port on which your TUIO is coming in if your going through a router (depending on your device) which we had to disable completely in order for it to work properly.

**Comment by spiral-labs** [comment]
Hi! thanks for the help, I got TUIO working. Now I'm trying to link a Tracker (Fiducial) to a game object so the position and rotation is applied to the game object.

Also, How can I enable the position information to show for every pointer, I've seen several videos that show the raw data, I enabled debugging but I cant get pointers to show this information.

**Comment by JosephOHagan** [comment]
@spiral-labs If it's what I think I setup a script to apply the inertia of the swipe input to the object.

Pointer? I think you might just have to dig around in the code where it sets up and interprets a touch point. 

You might have to convert from TUIO coordinates to something else throughout this process as well.

**Comment by spiral-labs** [comment]
If I could read the x,y,z and rotation info from the input pointer generated by the marker I would apply it to the game objects recttransform but I'm still not able to read sych data. Also I'm still not able to get the id that the marker represents so I can link it to a game object.


This is what I'm trying to achieve

In the video you can see Touchscript throws position data for each pointer, that is the info I'm not getting when touching my table.
https://www.youtube.com/watch?v=IgbMybbSfk4

**Comment by JosephOHagan** [comment]
@spiral-labs For my project I just hacked my code into the OnCursorAdded method of TuioInput.cs so you might have some lucking playing around with that class

One problem I did face was if you placed your index finger on the surface then your little finger it would assume that the index had simply moved to where the little finger was located. It might actually just think that was the new touch point to jump to. Not sure.

We got around that by limiting people to using 1 finger at a time (it was for a research experiment)

I've not had a chance to look into correcting it since then. 

From the video looks like it just plots the touch ID (which I think increments from 0 -> N +1 for every touch) and the tags saying what type of touch has occurred (there are 3 I think).

**Comment by spiral-labs** [comment]
Thank you @JosephOHagan 

Cursor Data is now showing, I had to create the Object Cursor prefab and assign it to the Cursor Manager, for some reason it was using a generic "Pointer" prefab for the Object Pointer Proxy.

Working with Tuioinput.cs is also working for me, only Instead of "OnCursorAdded" I focused on "OnObjectUpdated" since I'm working with markers. I'm trying to modify a GameObjects Transform with the information from the variable "touch.Position" but I get the error "get_transform can only be called from the main thread" do you know how can I fix this?

Thank you!

**Comment by sloopidoopi** [comment]
Hi,
you can't work with the callbacks from TuioInput because they are called not on the main thread which Unity API calls require. It's better to look into the CursorManager.cs   file (line 165) to get a picture which callbacks you have to use:   TouchManager.Instance.PointersUpdated for example.

Another way is to extend the ObjectCursor.cs file . To  rotate the ObjectCursors recttransform i added this to the class: 

```
public float deg
        {
            get;
            private set;
        }

protected override void update(IPointer pointer)
        {
            base.update(pointer);
            //Debug.Log(((ObjectPointer)pointer).Angle);           
            //Debug.Log(((ObjectPointer)pointer).Angle * Mathf.Rad2Deg);
            deg = ((ObjectPointer)pointer).Angle * Mathf.Rad2Deg;
            rect.rotation = Quaternion.Euler(0f, 0f, -deg);
        }
```
